textattack/transformations/word_swap_change_number.py

```diff
@@ -0,0 +1,161 @@
+from num2words import num2words
+import numpy as np
+from word2number import w2n
+
+from textattack.transformations import Transformation
+
+
+def idx_to_words(ls, words):
+    """Given a list generated from cluster_idx, return a list that contains
+    sub-list (the first element being the idx, and the second element being the
+    words corresponding to the idx)"""
+
+    output = []
+    for cluster in ls:
+        word = words[cluster[0]]
+        for idx in cluster[1:]:
+            word = " ".join([word, words[idx]])
+        output.append([cluster, word])
+    return output
+
+
+def cluster_idx(idx_ls):
+    """Given a list of idx, return a list that contains sub-lists of adjacent
+    idx."""
+
+    if len(idx_ls) < 2:
+        return [[i] for i in idx_ls]
+    else:
+        output = [[idx_ls[0]]]
+        prev = idx_ls[0]
+        list_pos = 0
+
+        for idx in idx_ls[1:]:
+            if idx - 1 == prev:
+                output[list_pos].append(idx)
+            else:
+                output.append([idx])
+                list_pos += 1
+            prev = idx
+        return output
+
+
+class WordSwapChangeNumber(Transformation):
+    def __init__(self, max_change=1, n=3, **kwargs):
+        """A transformation that recognize numbers in sentence, and return
+        sentences with altered numbers.
+
+        :param max_change: Maximum percent of change (1 being 100%)
+        :param n: Numbers of new numbers to generate
+        """
+        super().__init__(**kwargs)
+        self.max_change = max_change
+        self.n = n
+
+    def _get_transformations(self, current_text, indices_to_modify):
+        words = current_text.words
+        num_idx = []
+        num_words = []
+
+        # find indexes of alphabetical words
+        for idx in indices_to_modify:
+            word = words[idx].lower()
+            for number in STR_NUM:
```

Comment:
This seems slow -- O(WN), where W is the length of `indices_to_modify`, and N is the length of `STR_NUM`. We should try and avoid double loops like this. Any ideas on how to speed things up?

Reply:
I thought about doing something like:
```
for idx in indices_to_modify:

      word = words[idx].lower()

      if word in STR_NUM:

            ............................
```

But this will miss some of the number combinations like "thirty-five" since we don't have them in `STR_NUM`

Do you think this would be a worthy trade? Maybe there are better options we can think of.

Reply:
ok, understood – I think it's fine to leave as is for now 👍